### ISSUE TITLE
Optional Family Components

### ISSUE THREAD

**Issue by patternspandemic** [eliasku/ecx#9]
Greeting @eliasku !

I am having fun with ecx, and this is my first experience with such frameworks.

I've read that a `Family`'s components may be optional. Is this something that one must specify (if so how)? Or does the internal machinery gather into a `Family` all entities that may have one or more of the specified `Family` components?

~~Second, is it fair to assume that I need not create a `Family` when I'm looking to update entities in the context of a single component? I've seen in ecx-scene2d, that you simply `Wire` in the component and iterate over `_dirtyVector`. Is there much to gain over using a Family of a single component?~~ (Edit: Nevermind, read through the custom Transform component).

Thank you,
Brad

**Comment by eliasku** [comment]
Hi !

Yes, _dirtyVector is just hard-coded (for performance reason) event stream, with ENTITY_TRANSFORM_COULD_BE_DIRTY, each with entity ID data.

I want to do something like `?` in annotation of Family to mark component as optional. But for now just dont include these component to Family, and check them with component storage instance getted with `Wire`.

So for example you have:
`Family<Transform, Renderable> // ?Scissors component is optional`
then you just have
`Wire<Transform>`
`Wire<Renderable`
and 
`Wire<Scissors>`

Family guarantee that Transform and Renderable will be attached to each listed entity.
But you can check for each of them is Scissors component is also attached or not:

```
var scissors = _scissors.get(child);
if (scissors != null && !processScissors(entity)) {
    // entity fully clipped by scissors
    child = _node.__getNextSibling(child);
    continue;
}
```

**Comment by eliasku** [comment]
@patternspandemic also remember, `ecx` is in progress, and maybe not the best suitable ECS-framework for first experience :)
But I realised that you have got a lot, that's nice.

**Comment by eliasku** [comment]
I've added optional component notation:

```
var _entities:Family<Transform, Node, Renderable, (Scissors)>;
```

Scissors component are not required in example, but family declaration becomes more readable from some point...

**Comment by OvermindDL1** [comment]
Well I've been playing with ecx with Kha as the rendering side and it works wonderfully.  :-)

**Comment by patternspandemic** [comment]
> I want to do something like ? in annotation of Family to mark component as optional.

This was the first thing I tried, and figured it simply wasn't implemented yet.

> also remember, ecx is in progress, and maybe not the best suitable ECS-framework for first experience :)

Actually, I found the `Wire` capability allowed me to choose / experiment / switch between: easily getting things to work and having a more decoupled, 'pure' separation setup. `AutoComp` is great too, even if not being the first choice for performance. I am no where near having enough experience to worry about that :)

The optional component notation should help keep things clear, thanks.

> Well I've been playing with ecx with Kha as the rendering side and it works wonderfully. :-)

I'm using Kha as well, and found `Service` injection an easy way to pass around a canvas for systems to render to, as well as hook together Kha's scheduler and ecx-common's SystemRunner. Whether that's a great way to do it, not sure. But it works great for me so far!

Cool stuff!
